### PR TITLE
Refactored verifier for easier verifiers addition.

### PR DESCRIPTION
**Pull request by bwplotka**
Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: cmd/thanos/tools_bucket.go [hunk 1]
```diff
@@ -130,38 +126,32 @@ func registerBucketVerify(app extkingpin.AppClause, objStoreConfig *extflag.Path
...
 		if err != nil {
 			return err
 		}
 
-		if *repair {
-			v = verifier.NewWithRepair(logger, reg, bkt, backupBkt, fetcher, time.Duration(*deleteDelay), issues)
-		} else {
-			v = verifier.New(logger, reg, bkt, fetcher, time.Duration(*deleteDelay), issues)
-		}
+		ctx := context.Background()
```
Comment (kakkoyun) [inline cmd/thanos/tools_bucket.go:139]: nit: Shall we move closer to where it's first used?

File: pkg/verifier/verify.go [hunk 2]
```diff
@@ -18,85 +19,154 @@ import (
...
+type VerifierRepairer interface {
+	IssueID() string
+	VerifyRepair(ctx context.Context, conf Config, idMatcher func(ulid.ULID) bool, repair bool) error
 }
 
-func newVerifierMetrics(reg prometheus.Registerer) *verifierMetrics {
-	var m verifierMetrics
+// Config is an verifier config.
+type Config struct {
+	Logger      log.Logger
```
Comment (kakkoyun) [inline pkg/verifier/verify.go:34]: I think this shouldn't be part of the config semantically. We should pass logger and metrics directly, (I know it'd mess up the interface however it's more sound in that way) or rather in `Context` even though it's not safe enough. 

We have similar patterns in the code base, but this is something that disturbs me. We should find a better pattern do pass this type of global dependencies (GlobalContext?).

**Review by kakkoyun — COMMENTED** [review]
Looks good. I have added a couple of comments.

File: pkg/verifier/verify.go (hunk 2 above)
Comment (bwplotka) [reply pkg/verifier/verify.go:34]: I thought about context but it's also wrong ;p

Comment (bwplotka) [reply pkg/verifier/verify.go:34]: any suggestions?

**Comment by bwplotka** [comment]
Will fix in another PR (: Have something now on top of this :hugs:

File: pkg/verifier/verify.go (hunk 2 above)
Comment (kakkoyun) [reply pkg/verifier/verify.go:34]: In this particular implementation I'd suggest:
1. Not to log or increment in the implementations of the interfaces. Try to do that on the wrapper modules
2. Or just create another struct and pass that to the implementations. 

WDYT?

Comment (kakkoyun) [reply pkg/verifier/verify.go:34]: Or we could use constructors to pass down those dependencies.

**Pull request merged**